File: app/src/main/java/fr/nihilus/mymusic/database/Playlist.java
package fr.nihilus.mymusic.database;

import android.arch.persistence.room.ColumnInfo;
import android.arch.persistence.room.Entity;
import android.arch.persistence.room.Index;
import android.arch.persistence.room.PrimaryKey;
import android.net.Uri;
import android.support.annotation.Nullable;

import java.util.Date;

/**
 * A class that groups informations associated with a playlist.
 * Each playlist has an unique identifier that its {@link PlaylistTrack} children must reference
 * to be included.
 */
@Entity(tableName = "playlist",
        indices = {@Index(value = "name", unique = true)})
public class Playlist {

    @PrimaryKey(autoGenerate = true)
    @ColumnInfo(name = "id")
    private Long id;

    @ColumnInfo(name = "name")
    private String name;

    @ColumnInfo(name = "date_created")
    private Date created;

    @ColumnInfo(name = "date_last_played")
    private Date lastPlayed;

    @ColumnInfo(name = "art_uri")
    private Uri artUri;

    /**
     * Returns the unique identifier of this playlist.
     * If this playlist has not been saved, this identifier will be {@code null}.
     * @return unique identifier or {@code null}
     */
    @Nullable
    public Long getId() {
        return id;
    }

    void setId(Long id) {
        this.id = id;
    }

    /**
     * @return The name given by the user to this playlist
     */
    public String getName() {
        return name;
    }

    /**
     * Rename this playlist.
     * @param name The new name of this playlist
     */
    public void setName(String name) {
        this.name = name;
    }

    /**
     * @return the date at which this playlist has been created
     */
    public Date getCreated() {
        return created;
    }

    void setCreated(Date created) {
        this.created = created;
    }

    /**
     * @return the date at which this playlist has been last played
     */
    public Date getLastPlayed() {
        return lastPlayed;
    }

    /**
     * Update the date a which this playlist has been played for the last time.
     * @param lastPlayed date of last play
     */
    public void setLastPlayed(Date lastPlayed) {
        this.lastPlayed = lastPlayed;
    }

    /**
     * @return an URI pointing to a Bitmap featuring this playlist
     */
    public Uri getArtUri() {
        return artUri;
    }

    /**
     * Sets an URI pointing to a Bitmap that features this playlist.
     * @param artUri URI pointing to a Bitmap resource
     */
    public void setArtUri(Uri artUri) {
        this.artUri = artUri;
    }
}

File: media/src/main/java/fr/nihilus/music/media/playlists/Playlist.kt
/*
 * Copyright 2018 Thibault Seisel
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package fr.nihilus.music.media.playlists

import android.net.Uri
import android.support.v4.media.MediaDescriptionCompat
import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.Ignore
import androidx.room.PrimaryKey
import fr.nihilus.music.media.CATEGORY_PLAYLISTS
import fr.nihilus.music.media.mediaIdOf

/**
 * A class that groups information associated with a playlist.
 * Each playlist has an unique identifier that its [PlaylistTrack] children must reference
 * to be included.
 */
@Entity(tableName = "playlist")
internal data class Playlist(

    /**
     * The unique identifier of this playlist.
     * If this playlist has not been saved, this identifier will be `null`.
     */
    @PrimaryKey(autoGenerate = true)
    @ColumnInfo(name = "id")
    val id: Long?,

    /**
     * The title given by the user to this playlist.
     */
    @ColumnInfo(name = "title")
    val title: String,

    /**
     * The date at which this playlist has been created.
     */
    @ColumnInfo(name = "date_created")
    val created: Long,

    /**
     * Uri pointing to a Bitmap that illustrates this playlist.
     */
    @ColumnInfo(name = "icon_uri")
    val iconUri: Uri?
) {

    /**
     * Create a new unsaved playlist that has the given [title] and an optional icon.
     *
     * @param title The title of the new playlist.
     * @param iconUri The Uri pointing to an image file that represents the playlist.
     */
    @Ignore
    constructor(
        title: CharSequence,
        iconUri: Uri? = null
    ) : this(null, title.toString(), System.currentTimeMillis(), iconUri)

    fun asMediaDescription(builder: MediaDescriptionCompat.Builder): MediaDescriptionCompat {
        val playlistId = checkNotNull(id) { "Cant create MediaDescription of an unsaved playlist" }
        val mediaId = mediaIdOf(CATEGORY_PLAYLISTS, playlistId.toString())
        return builder.setMediaId(mediaId)
            .setTitle(title)
            .setIconUri(iconUri)
            .build()
    }
}
